src/core/qgsmaprendererjob.cpp

```diff
@@ -508,19 +706,105 @@ QImage QgsMapRendererJob::composeImage( const QgsMapSettings &settings, const La
   }
 
   painter.end();
+#if DEBUG_RENDERING
+  image.save( "/tmp/final.png" );
+#endif
   return image;
 }
 
-void QgsMapRendererJob::logRenderingTime( const LayerRenderJobs &jobs, const LabelRenderJob &labelJob )
+void QgsMapRendererJob::composeSecondPass( LayerRenderJobs &secondPassJobs, LabelRenderJob &labelJob )
+{
+#if DEBUG_RENDERING
+  int i = 0;
+#endif
+  // compose the second pass with the mask
+  for ( LayerRenderJob &job : secondPassJobs )
+  {
+#if DEBUG_RENDERING
+    i++;
+    job.img->save( QString( "/tmp/second_%1.png" ).arg( i ) );
+    int mask = 0;
+#endif
+
+    // Merge all mask images into the first one if we have more than one mask image
+    if ( job.maskPass.maskJobs.size() > 1 )
+    {
+      QPainter *maskPainter = nullptr;
+      for ( LayerRenderJob *maskJob : job.maskPass.maskJobs )
+      {
+        QImage *maskImage = maskJob ? maskJob->maskPass.image : labelJob.maskImage;
+#if DEBUG_RENDERING
+        maskImage->save( QString( "/tmp/mask_%1_%2.png" ).arg( i ).arg( mask++ ) );
+#endif
+        if ( ! maskPainter )
+        {
+          maskPainter = maskJob ? maskJob->context.maskPainter() : labelJob.context.maskPainter();
+        }
+        else
+        {
+          maskPainter->drawImage( 0, 0, *maskImage );
+        }
+      }
+    }
+
+    if ( ! job.maskPass.maskJobs.isEmpty() )
+    {
+      // All have been merged into the first
+      LayerRenderJob *maskJob = *job.maskPass.maskJobs.begin();
+
+      QImage *maskImage = maskJob ? maskJob->maskPass.image : labelJob.maskImage;
+#if DEBUG_RENDERING
+      maskImage->save( QString( "/tmp/mask_%1.png" ).arg( i ) );
+#endif
+
+      // Only retain parts of the second rendering that are "inside" the mask image
+      QPainter *painter = job.context.painter();
+      painter->setCompositionMode( QPainter::CompositionMode_DestinationIn );
+      painter->drawImage( 0, 0, *maskImage );
+#if DEBUG_RENDERING
+      job.img->save( QString( "/tmp/second_%1_a.png" ).arg( i ) );
+#endif
+
+      // Modify the first pass' image ...
+      QPainter *painter1 = job.maskPass.firstPassJob->context.painter();
```

Comment:
Hm - this approach makes masking incompatible with GeoPDF export, where we render individual layers to separate QPainters one-at-a-time, and the painters won't be available after each layer is rendered.

Just something to be aware of -- the two features are mutually exclusive.